libraries/AP_Math/SCurve.cpp

```diff
@@ -0,0 +1,963 @@
+        segment[SEG_INIT].jerk_ref = 0.0f;
+        segment[SEG_INIT].end_time = segment[SEG_ACCEL_END].end_time;
+        segment[SEG_INIT].end_accel = segment[SEG_ACCEL_END].end_accel;
+        segment[SEG_INIT].end_vel = segment[SEG_ACCEL_END].end_vel;
+        segment[SEG_INIT].end_pos = segment[SEG_ACCEL_END].end_pos;
+
+        // move change segments to acceleration segments
+        for (uint8_t i = SEG_INIT+1; i <= SEG_ACCEL_END; i++) {
+            segment[i].seg_type = segment[i+7].seg_type;
+            segment[i].jerk_ref = segment[i+7].jerk_ref;
+            segment[i].end_time = segment[i+7].end_time;
+            segment[i].end_accel = segment[i+7].end_accel;
+            segment[i].end_vel = segment[i+7].end_vel;
+            segment[i].end_pos = segment[i+7].end_pos;
+        }
+
+        // set change segments to last acceleration speed
+        for (uint8_t i = SEG_ACCEL_END+1; i <= SEG_CHANGE_END; i++) {
+            segment[i].seg_type = SegmentType::CONSTANT_JERK;
+            segment[i].jerk_ref = 0.0f;
+            segment[i].end_time = segment[SEG_ACCEL_END].end_time;
+            segment[i].end_accel = 0.0f;
+            segment[i].end_vel = segment[SEG_ACCEL_END].end_vel;
+            segment[i].end_pos = segment[SEG_ACCEL_END].end_pos;
+        }
+
+    } else if (time >= segment[SEG_CHANGE_END].end_time && time <= segment[SEG_CONST].end_time) {
+        // In the constant speed phase
+        // Move adjust phase to acceleration phase to provide room for further speed adjustments
+
+        // set initial segment to last acceleration segment
+        segment[SEG_INIT].seg_type = SegmentType::CONSTANT_JERK;
+        segment[SEG_INIT].jerk_ref = 0.0f;
+        segment[SEG_INIT].end_time = segment[SEG_CHANGE_END].end_time;
+        segment[SEG_INIT].end_accel = 0.0f;
+        segment[SEG_INIT].end_vel = segment[SEG_CHANGE_END].end_vel;
+        segment[SEG_INIT].end_pos = segment[SEG_CHANGE_END].end_pos;
+
+        // set acceleration and change segments to current constant speed
+        float Jt_out, At_out, Vt_out, Pt_out;
+        get_jerk_accel_vel_pos_at_time(time, Jt_out, At_out, Vt_out, Pt_out);
+        for (uint8_t i = SEG_INIT+1; i <= SEG_CHANGE_END; i++) {
+            segment[i].seg_type = SegmentType::CONSTANT_JERK;
+            segment[i].jerk_ref = 0.0f;
+            segment[i].end_time = time;
+            segment[i].end_accel = 0.0f;
+            segment[i].end_vel = Vt_out;
+            segment[i].end_pos = Pt_out;
+        }
+    }
+
+    // Adjust the INIT and ACCEL segments for new speed
+    if ((time <= segment[SEG_ACCEL_MAX].end_time) && is_positive(segment[SEG_ACCEL_MAX].end_time - segment[SEG_ACCEL_MAX-1].end_time) && (vel_max < segment[SEG_ACCEL_END].end_vel) && is_positive(segment[SEG_ACCEL_MAX].end_accel) ) {
+        // Path has not finished constant positive acceleration segment
+        // Reduce velocity as close to target velocity as possible
```

Comment:
My understanding is that users are unlikely to notice and I'm not sure what we would do about it even if we did know this was happening.  We know it will happen in some cases but I think we are handling it as best we can.